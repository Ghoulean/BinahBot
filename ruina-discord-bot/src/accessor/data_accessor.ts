import {
    DecoratedAbnoPage,
    Localization,
    LookupResult,
} from "@ghoulean/ruina-common";
import * as fastestLevenshtein from "fastest-levenshtein";
import * as __AUTOCOMPLETE from "../data/autocomplete.json";
import * as __CN_ABNO from "../data/cn/abno.json";
import * as __EN_ABNO from "../data/en/abno.json";
import * as __JP_ABNO from "../data/jp/abno.json";
import * as __KR_ABNO from "../data/kr/abno.json";
import * as __LOOKUP_RESULTS from "../data/queryLookupResults.json";
import * as __TRCN_ABNO from "../data/trcn/abno.json";

type QueryToLookupResult = {
    [key: string]: LookupResult[];
};

type QueryToDecoratedAbnopage = {
    [key: string]: DecoratedAbnoPage;
};

type LocalizationToQueryDecoratedAbnopage = {
    [key in Localization]: QueryToDecoratedAbnopage;
};

type LevenshteinResults = {
    best: string;
    score: number;
};

const LOOKUP_RESULTS: QueryToLookupResult =
    __LOOKUP_RESULTS as QueryToLookupResult;

const CN_ABNO: QueryToDecoratedAbnopage = __CN_ABNO as QueryToDecoratedAbnopage;
const EN_ABNO: QueryToDecoratedAbnopage = __EN_ABNO as QueryToDecoratedAbnopage;
const JP_ABNO: QueryToDecoratedAbnopage = __JP_ABNO as QueryToDecoratedAbnopage;
const KR_ABNO: QueryToDecoratedAbnopage = __KR_ABNO as QueryToDecoratedAbnopage;
const TRCN_ABNO: QueryToDecoratedAbnopage =
    __TRCN_ABNO as QueryToDecoratedAbnopage;

const LOCALIZATION_TO_DECORATED_ABNO_PAGE: LocalizationToQueryDecoratedAbnopage =
    {
        [Localization.CHINESE_SIMPLIFIED]: CN_ABNO,
        [Localization.CHINESE_TRADITIONAL]: TRCN_ABNO,
        [Localization.ENGLISH]: EN_ABNO,
        [Localization.JAPANESE]: JP_ABNO,
        [Localization.KOREAN]: KR_ABNO,
    };

const AUTOCOMPLETE: string[] = __AUTOCOMPLETE.data;

const FUZZY_MATCHING_DISTANCE = 2;

export class DataAccessor {
    constructor() {}

    public lookup(query: string, preferredLocale: Localization): LookupResult {
        let lookupResults: LookupResult[] | undefined = LOOKUP_RESULTS[query];
        if (!lookupResults) {
            const levenshteinResults: LevenshteinResults =
                this.levenshteinLookup(query);
            if (levenshteinResults.score > FUZZY_MATCHING_DISTANCE) {
                throw new Error(`Couldn't identify query result for ${query}.`);
            }
            lookupResults = LOOKUP_RESULTS[levenshteinResults.best];
        }
        for (const lookupResult of lookupResults) {
            if (lookupResult.locale == preferredLocale) {
                return lookupResult;
            }
        }
        return lookupResults[0];
    }

    public getDecoratedAbnoPage(
        pageId: string,
        locale: Localization
    ): DecoratedAbnoPage {
        const decoratedAbnoPage: DecoratedAbnoPage | undefined =
            this.getLocaledAbnoMapping(locale)[pageId];
        if (!decoratedAbnoPage) {
            throw new Error(`${pageId} not found in ${locale} locale`);
        }
        return decoratedAbnoPage;
    }

    public autocomplete(query: string): string[] {
        const retVal: string[] = [];
        const cleanQuery = this.cleanQuery(query);
        for (const q of AUTOCOMPLETE) {
            if (
                this.autocompleteDistance(cleanQuery, q) <=
                FUZZY_MATCHING_DISTANCE
            ) {
                retVal.push(q);
            }
        }
        return retVal.sort((a, b) => {
            return (
                this.autocompleteDistance(cleanQuery, a) -
                    this.autocompleteDistance(cleanQuery, b) ||
                a.length - b.length ||
                a.localeCompare(b)
            );
        });
    }

    private getLocaledAbnoMapping(
        locale: Localization
    ): QueryToDecoratedAbnopage {
        return LOCALIZATION_TO_DECORATED_ABNO_PAGE[locale];
    }

    private cleanQuery(s: string): string {
        return s.replace(/\s/g, "").toLowerCase();
    }

    private autocompleteDistance(query: string, lookup: string): number {
        return fastestLevenshtein.distance(
            query,
            this.cleanQuery(lookup).substring(0, query.length)
        );
    }

    private levenshteinLookup(s: string): LevenshteinResults {
        const closest: string = fastestLevenshtein.closest(s, AUTOCOMPLETE);
        return {
            best: closest,
            score: fastestLevenshtein.distance(
                this.cleanQuery(s),
                this.cleanQuery(closest)
            ),
        };
    }
}
