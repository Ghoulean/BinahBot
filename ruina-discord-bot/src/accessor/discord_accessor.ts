import {
    GetSecretValueCommandOutput,
    SecretsManager,
} from "@aws-sdk/client-secrets-manager";
import axios from "axios";
import { EnvVarRetriever } from "../util/env_var_retriever";

const BASE_DISCORD_URL: string = "https://discord.com/api/v10";
// TODO: reconsider where this value goes
// TODO: rename
const DISCORD_SECRET_ENV_KEY: string = "DISCORD_BOT_API_KEY_NAME";

// TODO: reconsider where this type should go
type DiscordSecretsValues = {
    applicationId: string;
    authToken: string;
};

export class DiscordAccessor {
    private readonly secretsManager: SecretsManager;
    private readonly envVarRetriever: EnvVarRetriever;
    private applicationId: string;
    private botAuthToken: string;

    constructor(
        secretsManager: SecretsManager,
        envVarRetriever: EnvVarRetriever
    ) {
        this.secretsManager = secretsManager;
        this.envVarRetriever = envVarRetriever;
    }

    public async writeResponse(
        interactionToken: string,
        // TODO: strongly typed embed
        embed: any
    ): Promise<void> {
        console.log(`Retrieving secrets (if not done so already)`);
        await this.retrieveSecrets();
        console.log(`Got secrets`);
        try {
            const discordPayload: any = {
                embeds: [embed],
            };
            console.log(
                `DiscordAccessor: writing response with applicationId=${this.applicationId} interactionToken=${interactionToken}`
            );
            console.log(
                `DiscordAccessor: writing response with discordPayload=${JSON.stringify(
                    discordPayload
                )}`
            );
            const { status } = await axios.post(
                `${BASE_DISCORD_URL}/webhooks/${this.applicationId}/${interactionToken}`,
                discordPayload,
                {
                    headers: {
                        Authorization: `Bot ${this.botAuthToken}`,
                        "Content-Type": "application/json",
                    },
                }
            );
            if (status === 204) {
                console.log("Successfully returned Discord response");
            }
        } catch (e) {
            throw e;
        }
    }

    private async retrieveSecrets(): Promise<void> {
        if (this.botAuthToken) {
            return;
        }
        const getSecretValueCommandOutput: GetSecretValueCommandOutput =
            await this.secretsManager.getSecretValue({
                SecretId: this.envVarRetriever.getRequired(
                    DISCORD_SECRET_ENV_KEY
                ),
            });
        if (getSecretValueCommandOutput.SecretString) {
            const secrets: DiscordSecretsValues = JSON.parse(
                getSecretValueCommandOutput.SecretString
            ) as DiscordSecretsValues;
            this.applicationId = secrets.applicationId;
            this.botAuthToken = secrets.authToken;
        }
    }
}
