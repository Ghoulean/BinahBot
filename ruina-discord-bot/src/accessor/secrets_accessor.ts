import {
    GetSecretValueCommand,
    GetSecretValueCommandInput,
    GetSecretValueCommandOutput,
    SecretsManager,
} from "@aws-sdk/client-secrets-manager";
import { DiscordSecrets } from "../model/discord/discord_secrets";

export class SecretsAccessor {
    private readonly secretsId: string;
    private readonly secretsManager: SecretsManager;
    private discordSecrets: DiscordSecrets;

    constructor(secretsId: string, secretsManager: SecretsManager) {
        this.secretsId = secretsId;
        this.secretsManager = secretsManager;
    }

    public async getDiscordSecrets(): Promise<DiscordSecrets> {
        return this.discordSecrets ?? await this.cacheDiscordSecrets();
    }

    private async cacheDiscordSecrets(): Promise<DiscordSecrets> {
        const input: GetSecretValueCommandInput = {
            SecretId: this.secretsId,
        };
        const command: GetSecretValueCommand = new GetSecretValueCommand(input);
        let discordApiKeys: GetSecretValueCommandOutput;
        try {
            discordApiKeys = await this.secretsManager.send(command);
        } catch (e: unknown) {
            throw new Error(`Unable to get Discord secrets: ${JSON.stringify(e)}`);
        }
        if (discordApiKeys.SecretString) {
            this.discordSecrets = JSON.parse(
                discordApiKeys.SecretString
            ) as DiscordSecrets;
        } else {
            throw new Error(
                `Got Discord secrets, but can't parse SecretString in ${JSON.stringify(discordApiKeys)}`
            );
        }
        return this.discordSecrets;
    }
}
