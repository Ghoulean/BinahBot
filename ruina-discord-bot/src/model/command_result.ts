export type CommandResult = {
    success: boolean;
    error?: string;
    payload?: any;
};