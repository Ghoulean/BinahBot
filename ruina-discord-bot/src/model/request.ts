import { Chapter, Localization } from "@ghoulean/ruina-common";

export type Request = {
    command: string;
    commandArgs: RequestCommandArgs;
    interactionToken: string;
    locale: Localization;
    chapter: Chapter;
    autocomplete?: boolean;
};

export type RequestCommandArgs = {
    [key: string]: string | number | boolean
};
