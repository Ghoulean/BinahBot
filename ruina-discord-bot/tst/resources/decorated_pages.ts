import {
    AbnoPageSelectType,
    Chapter,
    DecoratedAbnoPage,
    DecoratedCombatPage,
    DecoratedKeyPage,
    DecoratedPassive,
    DieType,
    Ego,
    Floor,
    Localization,
    Range,
    Rarity,
    Resistance,
} from "@ghoulean/ruina-common";

export const DECORATED_ABNO_PAGE: DecoratedAbnoPage = {
    locale: Localization.ENGLISH,
    name: "TestAbnoPage",
    description: "TestDescription",
    flavorText: "TestFlavorText",
    imagePath: "test.png",
    nameId: "test",
    floor: Floor.NONE,
    chapter: Chapter.LIBRARY_OF_RUINA,
    emoLevel: 1,
    emotionRate: 1,
    emotionSign: 1,
    targetType: AbnoPageSelectType.SELECT_ONE,
    scriptId: "testScriptId",
};

export const BASE_DECORATED_COMBAT_PAGE: DecoratedCombatPage = {
    locale: Localization.ENGLISH,
    name: "name",
    description: "description",
    diceDescriptions: ["description1"],
    id: "id",
    ego: Ego.NOT_EGO,
    egoFloor: Floor.BINAH,
    scriptId: "scriptId",
    rarity: Rarity.PAPERBACK,
    dice: [
        {
            type: DieType.BLUNT,
            min: 1,
            max: 2,
            scriptId: "scriptId",
        },
    ],
    chapter: Chapter.LIBRARY_OF_RUINA,
    cost: 0,
    range: Range.MELEE,
    imagePath: "image.png",
};

export const BASE_DECORATED_KEY_PAGE: DecoratedKeyPage = {
    name: "Keypage Name",
    passiveNames: ["passiveName1"],
    locale: Localization.ENGLISH,
    id: "pageId",
    textId: "textId",
    hp: 1,
    stagger: 2,
    hpResistances: {
        [DieType.SLASH]: Resistance.FATAL,
        [DieType.PIERCE]: Resistance.WEAK,
        [DieType.BLUNT]: Resistance.NORMAL
    },
    staggerResistances: {
        [DieType.SLASH]: Resistance.ENDURED,
        [DieType.PIERCE]: Resistance.INEFFECTIVE,
        [DieType.BLUNT]: Resistance.IMMUNE
    },
    minBaseSpeed: 1,
    maxBaseSpeed: 2,
    baseLight: 3,
    chapter: Chapter.LIBRARY_OF_RUINA,
    rarity: Rarity.PAPERBACK,
    defaultPassiveIds: ["passiveId1", "passiveId2"],
    exclusiveCardsIds: ["cardId1"]
}

export const BASE_DECORATED_PASSIVE: DecoratedPassive = {
    name: "Passive Name",
    description: "Description of passive",
    locale: Localization.ENGLISH,
    id: "passiveId",
    cost: 2,
    rarity: Rarity.PAPERBACK,
    canGivePassive: false
}