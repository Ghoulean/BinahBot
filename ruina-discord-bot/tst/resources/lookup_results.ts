import {
    Chapter,
    Localization,
    LookupResult,
    PageType,
} from "@ghoulean/ruina-common";
import { DisambiguationResults } from "../../src/model/disambiguation_result";

export const ABNO_LOOKUP_RESULT: LookupResult = {
    query: "query",
    chapter: Chapter.LIBRARY_OF_RUINA,
    locale: Localization.ENGLISH,
    pageType: PageType.ABNO_PAGE,
    pageId: "pageId",
};

export const COMBAT_LOOKUP_RESULT: LookupResult = {
    ...ABNO_LOOKUP_RESULT,
    pageType: PageType.COMBAT_PAGE,
};

export const KEYPAGE_LOOKUP_RESULT: LookupResult = {
    ...ABNO_LOOKUP_RESULT,
    pageType: PageType.KEY_PAGE,
};

export const PASSIVE_LOOKUP_RESULT: LookupResult = {
    ...ABNO_LOOKUP_RESULT,
    pageType: PageType.PASSIVE,
};

export const DISAMBIGUATION_LOOKUP_RESULT: LookupResult = {
    ...ABNO_LOOKUP_RESULT,
    pageType: PageType.DISAMBIGUATION,
};

export const DISAMBIGUATION_RESULTS: DisambiguationResults = {
    id: "disambiguation_id",
    locale: Localization.ENGLISH,
    query: "disambiguation",
    lookupResults: [ABNO_LOOKUP_RESULT, COMBAT_LOOKUP_RESULT],
};
