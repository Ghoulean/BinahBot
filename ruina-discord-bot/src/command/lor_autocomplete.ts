import { DataAccessor } from "../accessor/data_accessor";
import { CommandResult } from "../model/command_result";
import { Request } from "../model/request";

// TODO: think about where these constants go
const QUERY_COMMAND_ARG = "query";
const AUTOCOMPLETE_LIMIT = 5;

// TODO: consider interfacing autocomplete function
export class LorAutocomplete {
    private readonly dataAccessor: DataAccessor;

    constructor(dataAccessor: DataAccessor) {
        this.dataAccessor = dataAccessor;
    }

    public autocomplete(request: Request): CommandResult {
        const query: string = request.commandArgs[QUERY_COMMAND_ARG] as string;
        const autocompleteResults: string[] = this.dataAccessor
            .autocomplete(query)
            .slice(0, AUTOCOMPLETE_LIMIT);

        return {
            success: true,
            payload: autocompleteResults,
        };
    }
}
