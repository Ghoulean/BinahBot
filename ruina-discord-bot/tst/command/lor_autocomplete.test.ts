import { Chapter, Localization } from "@ghoulean/ruina-common";
import { DataAccessor } from "../../src/accessor/data_accessor";
import { LorAutocomplete } from "../../src/command/lor_autocomplete";
import { CommandResult } from "../../src/model/command_result";
import { Request } from "../../src/model/request";
import { ABNO_LOOKUP_RESULT } from "../resources/lookup_results";

const REQUEST: Request = {
    command: "lor",
    commandArgs: {
        query: "",
    },
    interactionToken: "",
    locale: Localization.ENGLISH,
    chapter: Chapter.IMPURITAS_CIVITATIS,
};

const LARGE_AUTOCOMPLETE_SEARCH: number = 15;

const AUTOCOMPLETE_RESULT: string[] = Array(LARGE_AUTOCOMPLETE_SEARCH).fill(
    "result"
);

const mockDataAccessor = new (<new () => DataAccessor>(
    DataAccessor
))() as jest.Mocked<DataAccessor>;

let lorAutocomplete: LorAutocomplete;

beforeEach(() => {
    lorAutocomplete = new LorAutocomplete(mockDataAccessor);
});

test("should return autocomplete results", () => {
    mockDataAccessor.autocomplete = jest.fn();
    mockDataAccessor.lookup = jest.fn();

    mockDataAccessor.autocomplete.mockReturnValueOnce(AUTOCOMPLETE_RESULT);
    mockDataAccessor.lookup.mockReturnValueOnce({
        ...ABNO_LOOKUP_RESULT,
        displayQuery: "displayQuery"
    }).mockReturnValue(ABNO_LOOKUP_RESULT);

    const commandResult: CommandResult = lorAutocomplete.autocomplete(REQUEST);

    expect(commandResult.success).toBe(true);
    expect((commandResult.payload as string[]).length).toBeLessThan(
        LARGE_AUTOCOMPLETE_SEARCH
    );
    expect(commandResult).toMatchSnapshot();
});
