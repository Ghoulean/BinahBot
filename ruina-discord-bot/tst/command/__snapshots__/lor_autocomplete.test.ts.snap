// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return autocomplete results 1`] = `
{
  "payload": [
    "result",
    "result",
    "result",
    "result",
    "result",
  ],
  "success": true,
}
`;
